clarity/src/vm/ast/parser_v2/lexer/mod.rs

```diff
@@ -0,0 +1,2207 @@
+                start_line,
+                start_column,
+                end_line: self.last_line as u32,
+                end_column: self.last_column as u32,
+            },
+            token,
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn read_tokens() {
+        let mut lexer = Lexer::new("", false).unwrap();
+        assert_eq!(lexer.read_token().unwrap().token, Token::Eof);
+        assert_eq!(lexer.diagnostics.len(), 0);
+
+        let mut lexer = Lexer::new(" ", false).unwrap();
+        assert_eq!(lexer.read_token().unwrap().token, Token::Whitespace);
+        assert_eq!(lexer.diagnostics.len(), 0);
+
+        let mut lexer = Lexer::new("\t", false).unwrap();
+        assert_eq!(lexer.read_token().unwrap().token, Token::Whitespace);
+        assert_eq!(lexer.diagnostics.len(), 0);
+
+        let mut lexer = Lexer::new("\n", false).unwrap();
+        assert_eq!(lexer.read_token().unwrap().token, Token::Whitespace);
+        assert_eq!(lexer.diagnostics.len(), 0);
+
+        let mut lexer = Lexer::new("\r", false).unwrap();
+        assert_eq!(lexer.read_token().unwrap().token, Token::Whitespace);
+        assert_eq!(lexer.diagnostics.len(), 2);
+        assert_eq!(lexer.diagnostics[0].e, LexerError::UnsupportedLineEnding);
+        assert_eq!(lexer.diagnostics[1].e, LexerError::EditorCRLFMode);
+
+        lexer = Lexer::new("(", false).unwrap();
+        assert_eq!(lexer.read_token().unwrap().token, Token::Lparen);
+        assert_eq!(lexer.diagnostics.len(), 0);
+
+        lexer = Lexer::new(").unwrap()", false).unwrap();
+        assert_eq!(lexer.read_token().unwrap().token, Token::Rparen);
+        assert_eq!(lexer.diagnostics.len(), 0);
+
+        lexer = Lexer::new("{", false).unwrap();
+        assert_eq!(lexer.read_token().unwrap().token, Token::Lbrace);
+        assert_eq!(lexer.diagnostics.len(), 0);
+
+        lexer = Lexer::new("}", false).unwrap();
+        assert_eq!(lexer.read_token().unwrap().token, Token::Rbrace);
+        assert_eq!(lexer.diagnostics.len(), 0);
+
+        lexer = Lexer::new(":", false).unwrap();
+        assert_eq!(lexer.read_token().unwrap().token, Token::Colon);
+        assert_eq!(lexer.diagnostics.len(), 0);
+
+        lexer = Lexer::new(",", false).unwrap();
+        assert_eq!(lexer.read_token().unwrap().token, Token::Comma);
+        assert_eq!(lexer.diagnostics.len(), 0);
+
+        lexer = Lexer::new(".", false).unwrap();
+        assert_eq!(lexer.read_token().unwrap().token, Token::Dot);
+        assert_eq!(lexer.diagnostics.len(), 0);
+
+        lexer = Lexer::new("123", false).unwrap();
+        assert_eq!(
+            lexer.read_token().unwrap().token,
+            Token::Int("123".to_string())
+        );
+        assert_eq!(lexer.diagnostics.len(), 0);
+
+        lexer = Lexer::new("0123", false).unwrap();
+        assert_eq!(
+            lexer.read_token().unwrap().token,
+            Token::Int("123".to_string())
```

Comment:
Wait, shouldn't this be `Token::Int("123".to_string())`?  Why does the lexer strip leading 0's?  Isn't that the parser's job?

Reply:
Yes, good catch. This should have changed with the switch making the lexer return strings for integers. Fixed in 4851c3880.